### PR TITLE
Erofs bloom filter

### PR DESCRIPTION
**Pull request by alexlarsson**
This adds the COMPAT_XATTR_FILTER feature flag, and computes h_name_filter in all the inode xattr header.

Fixes https://github.com/containers/composefs/issues/167

This is based on and tested against the code in the [dev branch of erofs.git](https://git.kernel.org/pub/scm/linux/kernel/git/xiang/erofs.git/log/?h=dev), so this is not upstream quite yet. However, this likely will go in unchanged.

@hsiangkao wanna have a quick look?

### CODE REVIEW

**Comment by alexlarsson** [comment]
NOTE: This is a format-breaking change, so its also updates the test checksums. We should merge this before freezing the format.

File: libcomposefs/erofs_fs.h [hunk 1]
```diff
@@ -127,11 +128,30 @@ static inline bool erofs_inode_is_data_compressed(unsigned int datamode)
...
+#define EROFS_INODE_LAYOUT_COMPACT	0
+/* 64-byte on-disk inode */
+#define EROFS_INODE_LAYOUT_EXTENDED	1
+
 struct erofs_inode_chunk_info {
 	__le16 format;		/* chunk blkbits, etc. */
 	__le16 reserved;
 };
 
+union erofs_inode_i_u {
```
Comment (cgwalters) [inline libcomposefs/erofs_fs.h:141]: Seems like some of this could be split out as a prep patch?

File: libcomposefs/lcfs-writer-erofs.c [hunk 2]
```diff
@@ -36,6 +36,121 @@
 #include <assert.h>
 #include <linux/fsverity.h>
 
+/* The xxh32 hash function is copied from xxhash with this copyright:
```
Comment (cgwalters) [inline libcomposefs/lcfs-writer-erofs.c:39]: From *where* precisely?  A git URL and exact digest would be useful.

Taking the copy from the Linux kernel and massaging it to compile may not actually be *too* hard...

File: libcomposefs/erofs_fs.h (hunk 1 above)
Comment (alexlarsson) [reply libcomposefs/erofs_fs.h:141]: Well, actually, it does tweak the code to work with some new naming in the new file, but its nice to merge those to have each commit build.

File: libcomposefs/lcfs-writer-erofs.c (hunk 2 above)
Comment (alexlarsson) [reply libcomposefs/lcfs-writer-erofs.c:39]: I took it from:
https://www.spinics.net/lists/kernel/msg4859620.html
Which is taken from the kernel, but initially was from 
static const uint32_t PRIME32_1 = 2654435761U;

Comment (alexlarsson) [reply libcomposefs/lcfs-writer-erofs.c:39]: https://github.com/torvalds/linux/blob/master/lib/xxhash.c

Comment (cgwalters) [reply libcomposefs/lcfs-writer-erofs.c:39]: OK makes sense, but I more meant let's add this data into the comment in the code, so someone knows where to look for updates.

**Comment by hsiangkao** [comment]
> @hsiangkao wanna have a quick look?

Sorry for a bit late reply since I'm busy in the last test for commits of the next cycle.
Overall it looks good to me, will upstream xattr bloom filter feature once the merge window opens.

**Comment by alexlarsson** [comment]
Added the link to the kernel source of xxh32()

File: libcomposefs/erofs_fs.h (hunk 1 above)
Comment (alexlarsson) [reply libcomposefs/erofs_fs.h:141]: The changes were proposed upstream today, so I think we can merge the whole thing soon.

**Comment by alexlarsson** [comment]
upstreaming is here: https://lists.ozlabs.org/pipermail/linux-erofs/2023-August/009427.html

**Comment by hsiangkao** [comment]
> upstreaming is here: https://lists.ozlabs.org/pipermail/linux-erofs/2023-August/009427.html

it already lands upstream.

**Review by giuseppe — APPROVED** [review]
LGTM

**Pull request merged**